docs/rfcs/0004-agent-flow.md

```diff
@@ -0,0 +1,162 @@
+
+Agent Flow refers to both the execution, configuration and visual configurator of data flow.
+
+The primary goals of Agent Flow are
+
+1. Allow users to more easily understand the impact of their configuration
+2. Allow users to collect metrics from all running integrations across a  set of agents without installing extra dependencies
+3. Allow users to run components based on a dynamic environment
+4. Allow developers to easily add components to the system
+5. Ensure the Agent still maintains high performance on all platforms that the Agent currently supports
+6. Flow configs must be easily machine-readable and machine-writable to support tooling, such as formatting and GUI
+
+Goals of this design document
+
+* Define high level concepts and goals
+* Define high level concepts of the execution path
+
+This document represents ideals and not technical implementation. 
+
+Non Goals of this design document
+
+* Define a technical implementation or configuration implementation
+
+# 2. Broad Solution Path
+
+Conversation around should the components be assembled via message passing, via expressions, or a hybrid approach.
+
+**Note: Consider all examples pseudoconfig**
+
+## 2.1 Expression Based
+
+Expression based is writing expressions that allow referencing other components streams/outputs/values and using them directly. 
+
+**Pros**
+
+* Easier to Implement, evaluating expressions can map directly to existing config structs
+* Components are more reusable, you can pass basic types arounds (string, int, bool)
+
+**Cons**
+* Harder for users to wire things together
+* Harder to build a GUI for
+
+
+## 2.2 Message Based
+
+Message based is where components have no knowledge of other components and information is passed strictly via input and output streams. 
+
+**Pros**
+
+* Easier for users to understand the dependencies between components
+* Easier to build a GUI for
+* Ability to use more configuration formats easier
+
+**Cons**
+
+* More time consuming to implement, existing integrations/items would need to be componentized
+* Larger type system needed
+* More structured to keep the amount of types down
+
+Messages would require a more rigid and well defined type structure. For instance for getting credentials from various sources and passing those credentials around we would want to avoid the following.
+
+* MySQLCredentials
+* RedisCredentials
+* RemoteWriteCredentials
+* MySQLCredentialsSourceS3
+* MySQLCredentialsSourceVault
+* MySQLCredentialsSourceConsul
+* RedisCredentialsSourceS3
+* RedisCredentialsSourceVault
+* RedisCredentialsSourceConsul
+* RemoteWriteCredentialsSourceS3
+* RemoteWriteCredentialsSourceVault
+* RemoteWriteCredentialsSourceConsul
+
+And instead have one component that can read from many sources and outputs a single `Credential Type` and its up the destination component to intepret that correctly. 
```

Comment:
I'm not sure this is possible, though. What would a single Credential type look like There's many different types of credentials; how would you distinguish between Basic Auth and Bearer Auth?

Reply:
Likely an enum or some other flag. May force the user into storing their credentials in a specific format.  

```
{
  "bearer-token" : "1234"
}
```

or
```
{
"basic-auth" 
  {
     "username":"1234"
      "password": "1234"
  }
}
```